File: Java/FileSystem/AbstractFileSystemNode.java

abstract class AbstractFileSystemNode implements FileSystemNode {

    protected AbstractFileSystemNode parent;
    protected String name;

    @Override
    public final FileSystemNode getParent() {
        return parent;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public final String getPath() {
        String resultPath = "";
        FileSystemNode parentGoUp = this;

        while (parentGoUp != null) {
            resultPath = parentGoUp.getName().concat(resultPath);
            parentGoUp = parentGoUp.getParent();
        }

        return resultPath;
    }

    public void setParent(AbstractFileSystemNode inParent) {
        this.parent = inParent;
    }

    public void setName(String inName) {
        this.name = inName;
    }

}

File: Kotlin/FileSystemOnKotlin/src/main/kotlin/AbstractFileSystemNode.kt
abstract class AbstractFileSystemNode : FileSystemNode {
    override var parent: AbstractFileSystemNode? = null
    override var name: String = ""

    final override fun getPath(): String {
        var parentGoUp = this as AbstractFileSystemNode?
        val listOfNames = mutableListOf<String>()

        while (parentGoUp != null) {
            listOfNames.add(parentGoUp.name)
            parentGoUp = parentGoUp.parent
        }

        return listOfNames.reversed().joinToString(separator = "")
    }
}
